EQL: add support for partial shard results

luigidellaquila opened the pull request:
Fixes: https://github.com/elastic/elasticsearch/issues/114918

Allow EQL to execute queries also when some shards are unavailable.
This adds a new parameter (both available as a query parameter and in the request body) , `allow_partial_search_results` (default `false`, that can be configured as a cluster setting `xpack.eql.default_allow_partial_results`), that enables this behavior.

If shard failures happen during the query, the execution only takes into consideration the available shards, and returns an additional "shard_failures" value in the response

```
{
   "query": "sequence ...",
   "allow_partial_search_results": true
}
```

```
{
    "is_partial": false,
    "is_running": false,
    "took": ...,
    "timed_out": false,
    "hits": {
        ...
     },
    "shard_failures": [
        {
            "shard": ...,
            "index": ...,
            "node": ...,
            "reason": {
                "type": "no_shard_available_action_exception",
                "reason": null
            }
        }
    ]
}
```

Sequence queries are particularly problematic, because their result does not depend on a single record, but by the correlation of multiple records, so the result of a query can be slightly incorrect in case of partial shards (consider the case where a missing event clause matches records that are in missing shards, it will return false positives).
For this reason, we also provide an additional parameter, `allow_partial_sequence_results` to fine-tune the behavior of sequence queries:
- `allow_partial_sequence_results=false` (default): in case of missing shards, sequence queries always return empty results, to avoid mistakes
- `allow_partial_sequence_results=true`: the sequences are calculated only on the available data, ie. on records present in available shards.

If `allow_partial_search_results=false` (default), then `allow_partial_sequence_results` has no effect.

TODO:
- [x] Implement request parameter and query logic
- [x] Implement response (`shard_failures`)
- [x] Allow override for the default value (eg. with something similar to `search.default_allow_partial_results`)
- [x] Docs

luigidellaquila committed:
EQL: add support for partial shard failures

luigidellaquila committed:
Add shard_failures to the response

luigidellaquila committed:
Docs + request param

luigidellaquila committed:
Add cluster setting

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila committed:
Randomize spec tests

luigidellaquila committed:
More tests

luigidellaquila committed:
Update docs/changelog/116388.yaml

elasticsearchmachine commented:
Hi @luigidellaquila, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

astefan reviewed (commented):
Thanks, @luigidellaquila. It looks promising from my point of view. On the technical side of things - a review from someone in the ES team who is more familiar with partial results would help greatly.

I would also look into some tests that use CCS. I recently learned that `skip_unavailable` setting for a CCS cluster is a flavor of "unavailable shards" situation.

Also, some tests with Security enabled would help as well.

luigidellaquila committed:
Fine tune sequences behavior using allow_partial_sequence_results

If true (default), the sequences are calculated on the available shards
If false, in case of shard failures, the query does not return any results, but does not throw errors either

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/eql/allow_partial_search_results' into eql/allow_partial_search_results

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila committed:
allow_partial_sequence_results=false by default

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila commented:
@yctercero if you have a chance, please have a look at this PR, it practically contains the logic we discussed off-line

yctercero commented:
I think this aligns with what we discussed. Where we'll be able to return partial results for non-sequence EQL queries and disallow partial results for sequence queries. 

I think @marshallmain is out this week, but maybe @approksiu @MikePaquette can confirm this is the desired behavior? Note that users will be allowed to return partial results for sequence queries if they turn on the flag.

luigidellaquila commented:
> Note that users will be allowed to return partial results for sequence queries if they turn on the flag.

To clarify, partial results for sequences means _full sequences_, but calculated only on the available data (ie. not pieces of sequences / sequences with a smaller number of events)

luigidellaquila committed:
Add CCS test

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila committed:
Fix license

luigidellaquila committed:
Refactor test

luigidellaquila committed:
Make CCS test more deterministic

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila committed:
Cleanup

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

javanna commented on x-pack/plugin/eql/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/eql/10_basic.yml:
```diff
@@ -478,3 +478,34 @@ setup:
+  - match: {hits.events.2.fields.@timestamp: ["1580906096000"]}
+  - match: {hits.events.2.fields.id: [123]}
+  - match: {hits.events.2.fields.valid: [true]}
+  - match: {hits.events.2.fields.day_of_week: ["Wednesday"]}
```
would be good to have a test with a simulated shard failure to see that it gets printed out in the response.

luigidellaquila replied:
Can we simulate shard failures in yaml tests? Do you have an example?

javanna replied:
You can certainly cause shard failures, for instance by querying multiple indices, out of which one returns an error for the query. Can be simple mapping related issues like a type conflict etc.

luigidellaquila replied:
I'll have to play with it a bit; probably mapping related problems are not the way, since EQL does quite some pre-analysis based on field_caps before running the search queries, so this kind of problems will result in an early validation exception.
In the IT I just shut down nodes and make shards unavailable, but I don't think I can do the same with yaml tests.

luigidellaquila replied:
Added a few more tests here, simulating shard failures with painless.

javanna commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchResponse.java:
```diff
@@ -84,11 +93,12 @@ private static final class Fields {
         parser.declareString(optionalConstructorArg(), ID);
         parser.declareBoolean(constructorArg(), IS_RUNNING);
         parser.declareBoolean(constructorArg(), IS_PARTIAL);
+        parser.declareObjectArray(optionalConstructorArg(), (p, c) -> ShardSearchFailure.EMPTY_ARRAY, SHARD_FAILURES); // TODO fix this
```
not sure about the TODO fix this at the end of the line

luigidellaquila replied:
sorry, just a leftover

javanna commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchResponse.java:
```diff
@@ -116,6 +128,19 @@ public EqlSearchResponse(StreamInput in) throws IOException {
+                shardFailures = new ShardSearchFailure[size];
+                for (int i = 0; i < shardFailures.length; i++) {
+                    shardFailures[i] = readShardSearchFailure(in);
+                }
```
could we have used readArray here or some of the primitives? that probably saves some lines of code. Same for the write side of it.

luigidellaquila replied:
Done

javanna commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java:
```diff
@@ -60,6 +60,20 @@ public class EqlPlugin extends Plugin implements ActionPlugin, CircuitBreakerPlu
+        false,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
```
Do we need a setting to tweak the default given it's a flag exposed at the request level?

luigidellaquila replied:
I'm not completely sure tbh (maybe @yctercero has opinions here).
My thinking is that it could make our life easier on existing deployments, where we have hundreds of EQL rules already configured and we don't want to change them one by one, but maybe this can be addressed at higher level (in Kibana?).

luigidellaquila replied:
Yes, all the defaults are to `false` now.
The discussion here was more about the real need for a cluster setting to change these defaults, since we can always set these values per query.

astefan replied:
I'm with @javanna here. If there is no need for this, let's not make it more complicated than it should be. If there will be a desire in the future for this, we can add it.

The basic principle of **"it's easy to put stuff in, harder to take it out"** (think about BWC) applies here.

Same principle for having this as both **request parameter** and **request body parameter**: it complicates the logic unnecessarily and confuses users with a lot of options to think and reason about. I realize that ES query DSL has this in the same way (both **request parameter** and **request body parameter**) and I am not against this that much, but unless there is an actual need and request I wouldn't confuse users.

yctercero replied:
Let's go with the less complicated option where we would handle it on the Kibana side - say an advanced setting or something of the like. 

cc @approksiu

javanna reviewed (commented):
I left a couple comments

approksiu commented:
Hi all, I have a question, if it would be possible to know if the results are based on full or partial available data? 
I want to know in case of setting allow_partial_search_results=true what kind of results were returned to determine if it makes sense to re-run the rule to get complete results.

luigidellaquila commented:
@approksiu with current implementation, in case of shard failures you'll have an additional `"shard_failures": [...]` in the response. If it's present, you can assume the results are from partial available data.

luigidellaquila committed:
Cleanup serialization code

approksiu commented:
Thanks! Sounds good to me!
The "shard_failures": [...] message would be present irregardless of a setting I assume?

luigidellaquila commented:
> The "shard_failures": [...] message would be present irregardless of a setting I assume?

It will always be present in case of shard failures, when `allow_partial_search_results=true`

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila commented:
> Thanks @luigidellaquila. Is my understanding correct that if allow_partial_search_results=false, the "shard_failures": [...] message would not be present?

If `allow_partial_search_results=false` (that is the default) the behavior will be exactly as it is today, ie. if there are no shard failures you get the normal result; if there are shard failures you get an error.

luigidellaquila committed:
More yaml tests

luigidellaquila committed:
Add toml tests and fix REST params

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila commented:
Thanks for your comments folks, I think I addressed all your observations. Please let me know if you have further concerns.

@astefan with my latest commit I added support for partial results also in TOML tests, it should give us a bit more flexibility, and it also covers scenarios with security enabled.

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

yctercero reviewed (approved):
LGTM - thank you so much for working on this addition. I think this now sets us up well on our end to decide how we want to expose this functionality on our end.

luigidellaquila commented:
@yctercero @approksiu thanks for the off-line conversations and for confirming that the PR covers your needs.

@astefan @javanna please let me know if you have further remarks of if we can proceed.

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/eql/allow_partial_search_results' into eql/allow_partial_search_results

astefan commented on x-pack/plugin/eql/qa/common/src/main/resources/test_failing_shards.toml:
```diff
@@ -0,0 +1,78 @@
+allow_partial_search_results = true
+allow_partial_sequence_results = true
+expected_event_ids  = [1, 2]
+expect_shard_failures = true
```
Can you add here more "failing shards" tests in other scenarios? (optional fields, samples, samples with optional fields, sequences with optional fields, sequences with missing events...)

astefan commented on x-pack/plugin/eql/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/eql/10_basic.yml:
```diff
@@ -478,3 +506,68 @@ setup:
           query: 'sequence with maxspan=10d [network where user == "ADMIN"] ![network where used == "SYSTEM"]'
   - match: { error.root_cause.0.type: "verification_exception" }
   - match: { error.root_cause.0.reason: "Found 1 problem\nline 1:75: Unknown column [used], did you mean [user]?" }
+
```
Or maybe in this or a new yml file to add the more varied scenarios for "failed shards" queries that I mentioned above.

luigidellaquila replied:
I added more cases to toml tests

astefan commented on x-pack/plugin/eql/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/eql/10_basic.yml:
```diff
@@ -478,3 +506,68 @@ setup:
+        body:
+          query: 'process where user == "SYSTEM" and day_of_week == "Monday"'
+          fields: [{"field":"@timestamp","format":"epoch_millis"},"id","valid","day_of_week"]
+          allow_partial_search_results: true
```
Vary the placement of the parameter as a request parameter as well, not only the body of the request.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sample/SampleIterator.java:
```diff
@@ -174,6 +183,14 @@ private void queryForCompositeAggPage(ActionListener<Payload> listener, final Sa
+    private void addShardFailures(SearchResponse r) {
+        if (r.getShardFailures() != null) {
+            for (ShardSearchFailure shardFailure : r.getShardFailures()) {
+                shardFailures.put(shardFailure.toString(), shardFailure); // TODO find a better way to deduplicate
```
Or limit the number of shard failures. How does ES deal with this scenario?

luigidellaquila replied:
ES returns all the failures, and I tend to agree with this approach (returning complete information).
ShardSearchFailure doesn't have an equals/hashcode implementation, and since it's widely used, I'm not sure it's something we want to introduce now.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/BasicQueryClient.java:
```diff
@@ -46,12 +46,14 @@ public class BasicQueryClient implements QueryClient {
         this.client = eqlSession.client();
         this.indices = cfg.indices();
         this.fetchFields = cfg.fetchFields();
+        this.allowPartialSearchResults = eqlSession.configuration().allowPartialSearchResults();
```
```suggestion
        this.allowPartialSearchResults = cfg.allowPartialSearchResults();
```

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -731,8 +755,8 @@ private void doPayload(ActionListener<Payload> listener) {
-        if (completed.isEmpty()) {
-            listener.onResponse(new EmptyPayload(Type.SEQUENCE, timeTook()));
+        if (completed.isEmpty() || (allowPartialSequenceResults == false && shardFailures.isEmpty() == false)) {
+            listener.onResponse(new EmptyPayload(Type.SEQUENCE, timeTook(), shardFailures.values().toArray(new ShardSearchFailure[0])));
```
Here you already know the size of the shardfailures, why not use it to its most optimal way:

```suggestion
            listener.onResponse(new EmptyPayload(Type.SEQUENCE, timeTook(), shardFailures.values().toArray(new ShardSearchFailure[shardFailures.size()])));
```

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -918,4 +948,13 @@ public KeyAndOrdinal next() {
         };
     }
+
+    private void addShardFailures(SearchResponse r) {
```
I've seen this exact method in SampleIterator as well. Maybe extract it somewhere and re-use it?

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/execution/sequence/PITFailureTests.java:
```diff
@@ -132,7 +134,15 @@ public void testHandlingPitFailure() {
+                null,
+                matcher,
+                Collections.emptyList(),
+                randomBoolean(),
```
Awesome 👍

astefan reviewed (commented):
I've left a first round of comments. It looks good.

In PartialSearchResultsIT or in toml files or in yml files I think we can do better in terms of covering more scenarios (involving optional fields for example). TumblingWindow specifically and SampleIterator are complex enough to trigger an ActionListener to return in different use cases.

luigidellaquila commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/BaseEqlSpecTestCase.java:
```diff
@@ -163,13 +186,34 @@ protected ObjectPath runQuery(String index, String query) throws Exception {
+            if (allowPartialResultsInBody) {
+                builder.field("allow_partial_search_results", String.valueOf(allowPartialSearchResults));
+                if (allowPartialSequenceResults != null) {
+                    builder.field("allow_partial_sequence_results", String.valueOf(allowPartialSequenceResults));
```
`allow_partial_sequece_results` alone (or with  `allow_partial_search_results=false`) has no effect.
There are randomized tests for this in [PartialSearchResultsIT](https://github.com/elastic/elasticsearch/pull/116388/files#diff-394483381f9a9290704dec509e32ba564499cf341a3b0ee536f7463537137aebR210) , but probably I can extend them a bit, randomly setting `allow_partial_search_results=false`

luigidellaquila commented on docs/reference/eql/eql-search-api.asciidoc:
```diff
@@ -88,6 +88,53 @@ request that targets only `bar*` still returns an error.
 +
 Defaults to `true`.
 
+`allow_partial_search_results`::
```
I think when _search docs mention partial results at the beginning of the page, they refer to something slightly different, ie. the first part of a response that is still being calculated. 
Search results can also be partial because of missing shards, that is the same we have in EQL, regardless of the request being sync or async.

This said, we definitely need some tests for async EQL queries with `allow_partial_search_results=true`. I'm adding them.

luigidellaquila commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/EqlMissingEventsSpecTestCase.java:
```diff
@@ -27,9 +27,23 @@ public EqlMissingEventsSpecTestCase(
         Integer size,
-        Integer maxSamplesPerKey
+        Integer maxSamplesPerKey,
+        Boolean allowPartialSearchResults,
```
There is some randomization already, but I'm making it more complete.

luigidellaquila committed:
More tests

luigidellaquila committed:
Refactoring

luigidellaquila committed:
More tests

luigidellaquila committed:
API spec + more tests

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

luigidellaquila committed:
Fix conflict

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -127,7 +133,10 @@ public TumblingWindow(
         SequenceMatcher matcher,
-        List<List<Attribute>> listOfKeys
+        List<List<Attribute>> listOfKeys,
+        boolean allowPartialSearchResults,
```
Given that [`allowPartialSequenceResults` depends on `allowPartialSearchResults` being `true`](https://github.com/elastic/elasticsearch/pull/116388/files#diff-efd7c22da9196006301656246c0b7bd07626fbed7107d8104f710fb8a115835eR117), I am surprised to see no checks here and `allowPartialSequenceResults` just being used. Can you point me to where we make sure that `allowPartialSequenceResults` is valid **if** `allowPartialSearchResults` has the right value?

luigidellaquila replied:
There is no specific check, because in case of partial shard failures the query would fail much earlier. The final result will be an error 400.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchRequest.java:
```diff
@@ -135,6 +141,13 @@ public EqlSearchRequest(StreamInput in) throws IOException {
         if (in.getTransportVersion().onOrAfter(TransportVersions.V_8_7_0)) {
             maxSamplesPerKey = in.readInt();
         }
+        if (in.getTransportVersion().onOrAfter(TransportVersions.EQL_ALLOW_PARTIAL_SEARCH_RESULTS)) {
```
Do we have any tests for serialization to make sure in a mixed cluster tests everything is ok from this point of view?

luigidellaquila replied:
BWC serialization is unit tested in `EqlSearchRequestTests`.
I'm adding some randomization to mixed-node tests

astefan reviewed (approved):
In general, this LGTM. I've left some comments around testing, mostly. And also one related to TumblingWindow where the two settings related to sequences are not clear to me where they are properly set.

luigidellaquila commented on test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -2432,7 +2432,7 @@ public static void afterClass() throws Exception {
      *  After the cluster is stopped, there are a few netty threads that can linger, so we make sure we don't leak any tasks on them.
      */
-    static void awaitGlobalNettyThreadsFinish() throws Exception {
+    public static void awaitGlobalNettyThreadsFinish() throws Exception {
```
It's a leftover, I don't need it anymore. Reverting

luigidellaquila commented on x-pack/plugin/eql/qa/common/src/main/resources/test_failing_shards.toml:
```diff
@@ -0,0 +1,173 @@
+query = 'process where serial_event_id == 1 or broken == 1'
+allow_partial_search_results = true
+expected_event_ids  = [1]
+expect_shard_failures = true
```
That's true, there is some intrinsic unpredictability with shard failures. I tried to keep it simple to make it as deterministic as possible.

luigidellaquila committed:
Refactor tests

luigidellaquila committed:
Merge branch 'main' into eql/allow_partial_search_results

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/118862)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/118862">](https://github.com/elastic/elasticsearch/pull/118862) |
